File: iPhone/BundleFileReader/BundleFileReader.m
/**
 * Bundle File Reader Plugin
 * Copyright (c) 2011 Tim Fischbach (github.com/tf)
 * MIT licensed
 */

#import "BundleFileReader.h"

@implementation BundleFileReader

/**
 * See bundle_file_reader.js for documentation.
 */
- (void)readResource:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options {
	NSString* callbackId = [arguments objectAtIndex:0];
	NSString* resource = [arguments objectAtIndex:1];
	NSString* type = [arguments objectAtIndex:2];
	NSString* dir = [arguments objectAtIndex:3];

	PluginResult* result = nil;
	NSString* jsString = nil;

	NSString* filePath = [[NSBundle mainBundle] pathForResource:resource ofType:type inDirectory:dir];

	if (filePath) {
		NSData* data =  [[NSData alloc] initWithContentsOfFile:filePath];
		NSString* strBuffer = nil;

		if (data) {
			strBuffer = [[NSString alloc] initWithBytes: [data bytes] length: [data length] encoding: NSUTF8StringEncoding];
		}
		else {
			strBuffer = [[NSString alloc] initWithString: @""];
		}

		result = [PluginResult resultWithStatus: PGCommandStatus_OK messageAsString: [ strBuffer stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding] ];
		jsString = [result toSuccessCallbackString:callbackId];

		[strBuffer release];
		[data release];
	}
	else {
		result = [PluginResult resultWithStatus:PGCommandStatus_OK messageAsString: @"Bundle resource not found." ];
		jsString = [result toErrorCallbackString:callbackId];
	}

	if (jsString) {
		[self writeJavascript: jsString];
	}
}

@end

File: iPhone/BundleFileReader/bundle_file_reader.js
/**
 * Bundle File Reader Plugin
 * Copyright (c) 2011 Tim Fischbach (github.com/tf)
 * MIT licensed
 */

/*global plugins, PhoneGap*/

/**
 * Read files from application bundle.
 */
plugins.BundleFileReader = (function() {
	function BundleFileReader() {}

	/**
	 * Reads a file from the application bundle. The first three
	 * parameters correspond directly to the parameters of NSBundle
	 * pathForResource:ofType:inDirectory.
	 *
	 * @param resource        {String} Basename of resource file.
	 * @param type            {String} File extension.
	 * @param dir             {String} Directory relative to bundle.
	 * @param successCallback {Function} Receives file contents as
	 *                        first parameter.
	 * @param errorCallback   {Function} Invoked if resource cannot
	 *                        be found.
	 */
	BundleFileReader.prototype.readResource = function(resource, type, dir, successCallback, errorCallback) {
		PhoneGap.exec(function(contents) {
		    successCallback(decodeURIComponent(contents));
		}, errorCallback, "BundleFileReader", "readResource", [resource, type, dir]);
	};

	PhoneGap.addConstructor(function()  {
		if(!window.plugins) {
			window.plugins = {};
		}
		plugins.bundleFileReader = new BundleFileReader();
	});

	return BundleFileReader;
}());